src/v/cluster/tests/leader_balancer_test.cc

```diff
@@ -10,55 +10,354 @@
  */
 #define BOOST_TEST_MODULE leader_balancer
 
+#include "absl/container/flat_hash_map.h"
 #include "cluster/scheduling/leader_balancer_greedy.h"
+#include "leader_balancer_test_utils.h"
+#include "model/metadata.h"
 
+#include <absl/container/flat_hash_set.h>
 #include <boost/test/unit_test.hpp>
 
+#include <algorithm>
+#include <iterator>
+#include <tuple>
 #include <vector>
 
-BOOST_AUTO_TEST_CASE(greedy) {
-    cluster::leader_balancer_strategy::index_type index;
+using index_type = cluster::leader_balancer_strategy::index_type;
+using gbs = cluster::greedy_balanced_shards;
+using reassignment = cluster::leader_balancer_strategy::reassignment;
 
-    // 10 nodes, 2 cores -> 20 shards
-    std::vector<model::broker_shard> shards;
-    for (auto n = 0U; n < 10; n++) {
-        for (auto s = 0U; s < 2; s++) {
-            model::broker_shard shard{model::node_id(n), s};
-            shards.push_back(shard);
-        }
-    }
+/**
+ * @brief Basic validity checks on a reassignment.
+ */
+static void check_valid(const index_type& index, const reassignment& movement) {
+    // from == to is not valid (error should not decrease)
+    BOOST_REQUIRE(movement.from != movement.to);
 
-    size_t replica = 0;
-    for (auto shard : shards) {
-        // 10 groups per shard
-        for (auto g = 0U; g < 10; g++) {
-            raft::group_id group(g);
-            // 3 replica per group
-            std::vector<model::broker_shard> replicas;
-            replicas.push_back(shard); // the "leader"
-            while (replicas.size() != 3) {
-                if (shards[replica % shards.size()] != shard) {
-                    replicas.push_back(shards[replica % shards.size()]);
-                }
-                ++replica;
-            }
-            index[shard][group] = std::move(replicas);
-        }
-    }
```

Comment:
The logic to create the mock index is shared between this test and the benchmark so has been factored out into a separate header.